piper/keystroke.py

```diff
@@ -0,0 +1,128 @@
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+import gi
+gi.require_version("Gtk", "3.0")
+from gi.repository import Gdk, GObject, Gtk
+
+
+class KeyStroke(GObject.Object):
+    """The KeyStroke object represents a keyboard shortcut as pressed by the
+    user in the key mapping's capture mode."""
+
+    # GDK uses an offset of 8 from the keycodes defined in linux/input.h as
+    # used by evdev.
+    _XORG_KEYCODE_OFFSET = 8
+
+    def __init__(self, **kwargs):
+        """Intantiates a new KeyStroke."""
+        GObject.Object.__init__(self, **kwargs)
+        self._accel = ""
+        self._keycode = 0
+        self._keyval = 0
+        self._modifiers = []
+
+    @GObject.Property(type=str)
+    def accel(self):
+        """The accelerator of the current keystroke, as generated by
+        Gtk.accelerator_name(keyval, mask)."""
+        return self._accel
+
+    def set_key(self, keycode, keyval):
+        """Sets the given keycode and keyval as the key of this keystroke.
+
+        @param keycode The raw key code as reported by GDK, as int.
+        @param keyval The key, as int. See the `gdk/gdkkeysyms.h` header file
+                      for a complete list of GDK key codes.
+        """
+        # TODO: is a validity check necessary here? In principle, all uses of
+        # this method come from ButtonDialog.on_key_press, so they should all be
+        # valid. Otherwise, what does a validity check look like?
+        self._keycode = keycode
+        self._keyval = keyval
+        self._update_accel()
+
+    def add_modifier(self, keycode, keyval):
+        """Adds the given modifier keycode and keyval to the list of modifiers
+        of this keystroke.
+
+        @param keycode The raw key code, as int.
+        @param keyval The key, as int. See the `gdk/gdkkeysyms.h` header file
+                      for a complete list of GDK key codes.
+        """
+        # TODO: is a validity check necessary here? In principle, all uses of
+        # this method come from ButtonDialog.on_key_press, so they should all be
+        # valid. Otherwise, what does a validity check look like?
+        self._modifiers.append((keycode, keyval))
+        self._update_accel()
+
+    def clear(self):
+        """Clears the current keystroke."""
+        self._keycode = 0
+        self._keyval = 0
+        self._modifiers = []
+        self._update_accel()
+
+    def from_evdev(self, keycode, modifiers):
+        """Converts the given evdev keycodes to their GDK counterparts and sets
+        them as the current keystroke.
+
+        @param keycode The evdev keycode, as int.
+        @param modifiers A list of modifier keycodes, as [int].
+        """
+        # TODO: get keyval from keycode. Without this, the accelerator won't be
+        # printed correctly.
```

Comment:
you have to go through XKB but tbh I don't know exactly how this is handled here.  `gdk_keymap_translate_keyboard_state ()` seems like it's a good start, but you'll run into the various modifier issues here. 

I think the best option here may be to use `gdk_keymap_get_entries_for_keycode ()` and always select the entry on level 0. That gives you a key you can then feed into `gdk_keymap_lookup_key ()` to get the keyval, which has a printable name with `gdk_keyval_name()`

Together with a few warnings that we're recording keys not *syms* this may be enough (at least for now). It just means that when you're recording "shift 1" you don't record "!" but really "shift 1". At least using the get_entries call should provide basic layout mapping, so we don't have a confusion when someone uses azerty.